File: legacy/chaoxing_app/lib/models/course_catalog.dart
class CourseSpace {
  const CourseSpace({
    required this.courseId,
    required this.classId,
    required this.cpi,
    required this.title,
  });

  final String courseId;
  final String classId;
  final String cpi;
  final String title;

  String get key => '$courseId:$classId';

  factory CourseSpace.fromJson(Map<String, dynamic> json) {
    return CourseSpace(
      courseId: json['courseId'] is String ? json['courseId'] as String : '',
      classId: json['classId'] is String ? json['classId'] as String : '',
      cpi: json['cpi'] is String ? json['cpi'] as String : '',
      title: json['title'] is String ? json['title'] as String : '',
    );
  }

  Map<String, dynamic> toJson() => {
    'courseId': courseId,
    'classId': classId,
    'cpi': cpi,
    'title': title,
  };
}

class CoursePreference {
  const CoursePreference({required this.course, this.monitored = true});

  final CourseSpace course;
  final bool monitored;

  CoursePreference copyWith({CourseSpace? course, bool? monitored}) {
    return CoursePreference(
      course: course ?? this.course,
      monitored: monitored ?? this.monitored,
    );
  }

  factory CoursePreference.fromJson(Map<String, dynamic> json) {
    final rawCourse = json['course'];
    return CoursePreference(
      course: CourseSpace.fromJson(
        rawCourse is Map
            ? rawCourse.cast<String, dynamic>()
            : const <String, dynamic>{},
      ),
      monitored: json['monitored'] != false,
    );
  }

  Map<String, dynamic> toJson() => {
    'course': course.toJson(),
    'monitored': monitored,
  };
}

class CourseCatalog {
  const CourseCatalog({this.courses = const [], this.lastDiscoveredAt});

  final List<CoursePreference> courses;
  final DateTime? lastDiscoveredAt;

  static const empty = CourseCatalog();

  bool get hasRecords => courses.isNotEmpty;

  List<CourseSpace> get monitoredCourses => courses
      .where((preference) => preference.monitored)
      .map((preference) => preference.course)
      .toList();

  CourseCatalog mergeDiscovered(
    List<CourseSpace> discovered, {
    required DateTime discoveredAt,
  }) {
    final existing = {for (final value in courses) value.course.key: value};
    final merged = <CoursePreference>[];
    final discoveredKeys = <String>{};
    for (final course in discovered) {
      if (course.courseId.isEmpty || course.classId.isEmpty) {
        continue;
      }
      discoveredKeys.add(course.key);
      merged.add(
        CoursePreference(
          course: course,
          monitored: existing[course.key]?.monitored ?? true,
        ),
      );
    }
    merged.addAll(
      courses.where((value) => !discoveredKeys.contains(value.course.key)),
    );
    return CourseCatalog(courses: merged, lastDiscoveredAt: discoveredAt);
  }

  CourseCatalog setMonitored(String courseKey, bool monitored) {
    return CourseCatalog(
      courses: courses
          .map(
            (value) => value.course.key == courseKey
                ? value.copyWith(monitored: monitored)
                : value,
          )
          .toList(),
      lastDiscoveredAt: lastDiscoveredAt,
    );
  }

  factory CourseCatalog.fromJson(Map<String, dynamic> json) {
    final rawCourses = json['courses'];
    return CourseCatalog(
      courses: rawCourses is List
          ? rawCourses
                .whereType<Map>()
                .map(
                  (value) =>
                      CoursePreference.fromJson(value.cast<String, dynamic>()),
                )
                .where(
                  (value) =>
                      value.course.courseId.isNotEmpty &&
                      value.course.classId.isNotEmpty,
                )
                .toList()
          : const [],
      lastDiscoveredAt: DateTime.tryParse(
        json['lastDiscoveredAt'] is String
            ? json['lastDiscoveredAt'] as String
            : '',
      ),
    );
  }

  Map<String, dynamic> toJson() => {
    'courses': courses.map((value) => value.toJson()).toList(),
    if (lastDiscoveredAt != null)
      'lastDiscoveredAt': lastDiscoveredAt!.toIso8601String(),
  };
}

File: packages/chaoxing-domain/src/course-catalog.ts
import { asRecord } from "./json";

export type CourseSpace = {
  courseId: string;
  classId: string;
  cpi: string;
  title: string;
};

export type CoursePreference = {
  course: CourseSpace;
  monitored: boolean;
};

export type CourseCatalog = {
  courses: readonly CoursePreference[];
  lastDiscoveredAt: Date | null;
};

export const emptyCourseCatalog: CourseCatalog = {
  courses: [],
  lastDiscoveredAt: null,
};

export function courseSpaceKey(course: CourseSpace): string {
  return `${course.courseId}:${course.classId}`;
}

export function courseSpaceFromJson(json: Record<string, unknown>): CourseSpace {
  return {
    courseId: typeof json.courseId === "string" ? json.courseId : "",
    classId: typeof json.classId === "string" ? json.classId : "",
    cpi: typeof json.cpi === "string" ? json.cpi : "",
    title: typeof json.title === "string" ? json.title : "",
  };
}

export function courseSpaceToJson(course: CourseSpace): Record<string, string> {
  return {
    courseId: course.courseId,
    classId: course.classId,
    cpi: course.cpi,
    title: course.title,
  };
}

export function coursePreferenceFromJson(
  json: Record<string, unknown>,
): CoursePreference {
  return {
    course: courseSpaceFromJson(asRecord(json.course) ?? {}),
    monitored: json.monitored !== false,
  };
}

export function coursePreferenceToJson(
  preference: CoursePreference,
): Record<string, unknown> {
  return {
    course: courseSpaceToJson(preference.course),
    monitored: preference.monitored,
  };
}

export function catalogHasRecords(catalog: CourseCatalog): boolean {
  return catalog.courses.length > 0;
}

export function monitoredCourses(catalog: CourseCatalog): CourseSpace[] {
  return catalog.courses
    .filter((preference) => preference.monitored)
    .map((preference) => preference.course);
}

export function mergeDiscoveredCourses(
  catalog: CourseCatalog,
  discovered: readonly CourseSpace[],
  discoveredAt: Date,
): CourseCatalog {
  const existing = new Map(
    catalog.courses.map((value) => [courseSpaceKey(value.course), value]),
  );
  const merged: CoursePreference[] = [];
  const discoveredKeys = new Set<string>();
  for (const course of discovered) {
    if (course.courseId.length === 0 || course.classId.length === 0) {
      continue;
    }
    const key = courseSpaceKey(course);
    discoveredKeys.add(key);
    merged.push({
      course,
      monitored: existing.get(key)?.monitored ?? true,
    });
  }
  merged.push(
    ...catalog.courses.filter(
      (value) => !discoveredKeys.has(courseSpaceKey(value.course)),
    ),
  );
  return { courses: merged, lastDiscoveredAt: discoveredAt };
}

export function setCourseMonitored(
  catalog: CourseCatalog,
  courseKey: string,
  monitored: boolean,
): CourseCatalog {
  return {
    courses: catalog.courses.map((value) =>
      courseSpaceKey(value.course) === courseKey
        ? { ...value, monitored }
        : value,
    ),
    lastDiscoveredAt: catalog.lastDiscoveredAt,
  };
}

export function isSameLocalDate(left: Date, right: Date): boolean {
  return (
    left.getFullYear() === right.getFullYear() &&
    left.getMonth() === right.getMonth() &&
    left.getDate() === right.getDate()
  );
}

export function shouldDiscoverCourses(input: {
  catalog: CourseCatalog;
  now: Date;
  forceDiscovery: boolean;
}): boolean {
  return (
    input.forceDiscovery ||
    !catalogHasRecords(input.catalog) ||
    input.catalog.lastDiscoveredAt === null ||
    !isSameLocalDate(input.catalog.lastDiscoveredAt, input.now)
  );
}

export function courseCatalogFromJson(
  json: Record<string, unknown>,
): CourseCatalog {
  const rawCourses = json.courses;
  return {
    courses: Array.isArray(rawCourses)
      ? rawCourses
          .map((value) => asRecord(value))
          .filter((value): value is Record<string, unknown> => value !== null)
          .map(coursePreferenceFromJson)
          .filter(
            (value) =>
              value.course.courseId.length > 0 &&
              value.course.classId.length > 0,
          )
      : [],
    lastDiscoveredAt:
      typeof json.lastDiscoveredAt === "string"
        ? (() => {
            const parsed = new Date(json.lastDiscoveredAt);
            return Number.isNaN(parsed.getTime()) ? null : parsed;
          })()
        : null,
  };
}

export function courseCatalogToJson(
  catalog: CourseCatalog,
): Record<string, unknown> {
  return {
    courses: catalog.courses.map(coursePreferenceToJson),
    ...(catalog.lastDiscoveredAt
      ? { lastDiscoveredAt: catalog.lastDiscoveredAt.toISOString() }
      : {}),
  };
}
